### PR TITLE
GIDS: Update Data Requirements Display#5752

### PR DESCRIPTION
**Pull request by ksantiagoBAH**
## Description
As an EDU Team Member, I need it to be clear what the data requirements are for each file when uploading so that I can address any reported issues without having to navigate to multiple pages.

https://app.zenhub.com/workspaces/vft-59c95ae5fda7577a9b3184f8/issues/department-of-veterans-affairs/va.gov-team/5752

## Testing done
Local Testing Completed

## Acceptance criteria
- [x] The formatting of the data requirements help section is formatted using order lists and bullets (see Supporting Artifact # 1 for reference).
- [x] The data requirements help section is displayed on the file upload success page (see Supporting Artifact # 2 for reference).

## Definition of done
- [x] Events are logged appropriately
- [x] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: app/controllers/uploads_controller.rb [hunk 1]
```diff
@@ -9,42 +9,48 @@ def new
...
+      csv_requirements
+
       return
     end
-
     alert_and_log(@upload.errors.full_messages.join(', '))
     redirect_to dashboards_path
   end
 
+  def csv_requirements
```
Comment (zurbergram) [inline app/controllers/uploads_controller.rb]: can you move this method down below the `private` ?

File: app/controllers/uploads_controller.rb [hunk 2]
```diff
@@ -9,42 +9,48 @@ def new
...
-      @requirements = requirements_messages if @upload.csv_type_check?
       render :new
-    end
+   end
   end
 
   def show
     @upload = Upload.find_by(id: params[:id])
+
+    csv_requirements if @upload
```
Comment (zurbergram) [inline app/controllers/uploads_controller.rb]:
```suggestion
    csv_requirements if @upload.present?
```

File: app/controllers/uploads_controller.rb [hunk 3]
```diff
@@ -97,32 +103,75 @@ def klass
...
+  def batch
+    custom_batch_validator_messages = {
+      message: 'Requirement Description:',
+      value: "#{klass.name}Validator::REQUIREMENT_DESCRIPTIONS".safe_constantize
+    }
+    if custom_batch_validator_messages[:value].nil?
+      {}
+    else
+      [custom_batch_validator_messages]
+    end
```
Comment (zurbergram) [inline app/controllers/uploads_controller.rb]: Change the if/else block to
```suggestion
    [custom_batch_validator_messages] if custom_batch_validator_messages[:value].present?
```

File: app/controllers/uploads_controller.rb [hunk 4]
```diff
@@ -97,32 +103,75 @@ def klass
...
+  end
+
+  def validation_messages_inclusion
+    inclusion = []
+
     klass.validators.map do |validations|
       case validations
-      when ActiveRecord::Validations::PresenceValidator
-        generic_requirement_message('These columns must have a value: ', validations)
       when ActiveModel::Validations::InclusionValidator
```
Comment (zurbergram) [inline app/controllers/uploads_controller.rb]: change this to an if statement since only one case now

File: app/controllers/uploads_controller.rb [hunk 5]
```diff
@@ -97,32 +103,75 @@ def klass
...
+        array = { message: affected_attributes(validations), value: inclusion_requirement_message(validations) }
+        inclusion.push(array)
       end
     end
+
+    if !inclusion.empty?
+      [inclusion]
+    else
+      inclusion = {}
+    end
```
Comment (zurbergram) [inline app/controllers/uploads_controller.rb]: Change the if/else block to

```suggestion
    [inclusion] unless inclusion.empty?
```

File: app/controllers/uploads_controller.rb [hunk 6]
```diff
@@ -97,32 +103,75 @@ def klass
...
+
+    return [numericality] unless numericality[:value].empty?
+  end
+
+  def validation_messages_uniqueness
+    uniqueness = { message: 'These columns should contain unique values: ', value: [] }
+
+    uniqueness[:value] = klass_validator(ActiveRecord::Validations::UniquenessValidator)
+
+    return [uniqueness] unless uniqueness[:value].empty?
```
Comment (zurbergram) [inline app/controllers/uploads_controller.rb]:
```suggestion
   [uniqueness] unless uniqueness[:value].empty?
```

File: app/controllers/uploads_controller.rb [hunk 7]
```diff
@@ -97,32 +103,75 @@ def klass
...
+
+    return [presence] unless presence[:value].empty?
+  end
+
+  def validation_messages_numericality
+    numericality = { message: 'These columns can only contain numeric values: ', value: [] }
+
+    numericality[:value] = klass_validator(ActiveModel::Validations::NumericalityValidator)
+
+    return [numericality] unless numericality[:value].empty?
```
Comment (zurbergram) [inline app/controllers/uploads_controller.rb]:
```suggestion
   [numericality] unless numericality[:value].empty?
```

File: app/controllers/uploads_controller.rb (hunk 1 above)
Comment (ksantiagomorgan6) [reply app/controllers/uploads_controller.rb]: done

File: app/models/program.rb [hunk 8]
```diff
@@ -14,6 +14,7 @@ class Program < ApplicationRecord
     'length' => { column: :length, converter: BaseConverter }
   }.freeze
 
-  validates :facility_code, :description, presence: true
+  validates :facility_code, presence: true
+  validates :description, presence: true
```
Comment (zurbergram) [inline app/models/program.rb]: do these need to be on a separate line?

File: app/models/upload.rb [hunk 9]
```diff
@@ -116,9 +116,14 @@ def self.fetching_for?(csv_type)
     Upload.where(ok: false, completed_at: nil, csv_type: csv_type).any?
   end
 
+  # def self.valid_col_seps
```
Comment (zurbergram) [inline app/models/upload.rb]: remove the commented out code

File: app/views/uploads/new.html.erb [hunk 10]
```diff
@@ -9,17 +9,52 @@
...
       <%= render 'form' %>
     </div>
-    <% if @requirements&.any? %>
-      <div class="col-xs-4">
-        <div class="alert alert-info" role="alert">
-          <h4 class="alert-heading">CSV Data Requirements</h4><br />
-          <ul>
-            <% @requirements.each do |validations| %>
-              <li><%= validations %></li>
+        <% if @requirements&.any? %>
```
Comment (zurbergram) [inline app/views/uploads/new.html.erb]: Should this be broken out into a partial since it is displayed in both new and show?

File: app/views/uploads/show.html.erb [hunk 11]
```diff
@@ -66,3 +66,54 @@
     <%= @upload.comment %>
   </div>
 </div>
+
+<div class="container" style="float: left;">
```
Comment (zurbergram) [inline app/views/uploads/show.html.erb]: Should this be broken out into a partial since it is displayed in both new and show?

File: spec/controllers/uploads_controller_spec.rb [hunk 12]
```diff
@@ -79,28 +79,39 @@ def map_attributes(csv_class, requirement_class)
       requirements(csv_class, requirement_class)
         .attributes
         .map { |column| csv_class::CSV_CONVERTER_INFO.select { |_k, v| v[:column] == column }.keys.join(', ') }
-        .join(', ')
     end
 
+    # def map_attributes(csv_class, requirement_class)
```
Comment (zurbergram) [inline spec/controllers/uploads_controller_spec.rb]: removed commented out code

File: app/models/upload.rb [hunk 13]
```diff
@@ -116,9 +116,14 @@ def self.fetching_for?(csv_type)
...
+  #   valid_col_seps = Settings.csv_upload.column_separators.map { |cs| "\"#{cs}\"" }.join(' and ')
+  #   "Valid column separators are: #{valid_col_seps}."
+  # end
+
   def self.valid_col_seps
```
Comment (zurbergram) [inline app/models/upload.rb]: how does changing this method affect the usage in the method `Upload.set_col_sep` down below in this file?

File: app/models/program.rb (hunk 8 above)
Comment (ksantiagomorgan6) [reply app/models/program.rb]: Yes there are quite a few like this, i'll fix all the models

File: app/controllers/uploads_controller.rb [hunk 14]
```diff
@@ -131,12 +181,11 @@ def csv_column_name(column)
     klass::CSV_CONVERTER_INFO.select { |_k, v| v[:column] == column }.keys.join(', ')
   end
 
-  def generic_requirement_message(message, validations)
-    message + affected_attributes(validations)
+  def generic_requirement_message(validations)
```
Comment (zurbergram) [inline app/controllers/uploads_controller.rb]: this method isn't needed since it just calls `affected_attributes(validations)`

File: app/models/hcm.rb [hunk 15]
```diff
@@ -14,7 +14,9 @@ class Hcm < ApplicationRecord
     'method reason desc' => { column: :hcm_reason, converter: BaseConverter }
   }.freeze
 
-  validates :ope, :hcm_type, :hcm_reason, presence: true
+  validates :ope, presence: true
```
Comment (zurbergram) [inline app/models/hcm.rb]: undo these changes

File: app/models/ipeds_ic.rb [hunk 16]
```diff
@@ -121,7 +121,10 @@ class IpedsIc < ApplicationRecord
   }.freeze
 
   validates :cross, presence: true
-  validates :vet2, :vet3, :vet4, :vet5, inclusion: { in: (-2..1) }
+  validates :vet2, inclusion: { in: (-2..1) }
```
Comment (zurbergram) [inline app/models/ipeds_ic.rb]: undo these changes

File: app/models/mou.rb [hunk 17]
```diff
@@ -16,7 +16,8 @@ class Mou < ApplicationRecord
     'approval date' => { column: :approval_date, converter: BaseConverter }
   }.freeze
 
-  validates :ope, :ope6, presence: true
+  validates :ope, presence: true
```
Comment (zurbergram) [inline app/models/mou.rb]: undo do these changes

File: app/models/program.rb [hunk 18]
```diff
@@ -14,6 +14,7 @@ class Program < ApplicationRecord
     'length' => { column: :length, converter: BaseConverter }
   }.freeze
 
-  validates :facility_code, :description, presence: true
+  validates :facility_code, presence: true
```
Comment (zurbergram) [inline app/models/program.rb]: undo do these changes

File: app/models/settlement.rb [hunk 19]
```diff
@@ -14,4 +14,5 @@ class Settlement < ApplicationRecord
   }.freeze
 
   validates :cross, :settlement_description, presence: true
+  validates :settlement_description, presence: true
```
Comment (zurbergram) [inline app/models/settlement.rb]: undo do these changes

File: app/models/weam.rb [hunk 20]
```diff
@@ -74,7 +74,9 @@ class Weam < ApplicationRecord
   }.freeze
 
   has_many :crosswalk_issue, dependent: :delete_all
-  validates :facility_code, :institution, :country, presence: true
+  validates :facility_code, presence: true
```
Comment (zurbergram) [inline app/models/weam.rb]: undo do these changes

**Review by zurbergram — CHANGES_REQUESTED** [review]
Please undo the changes to all the models and instead update how the method `affected_attributes` works which might require a tweak to how `validation_messages_inclusion` works as well

**Review by zurbergram — APPROVED** [review]

**Review by sonntag-adam — APPROVED** [review]

File: app/controllers/uploads_controller.rb [hunk 21]
```diff
@@ -97,46 +103,83 @@ def klass
...
+  def batch
+    custom_batch_validator_messages = {
+      message: 'Requirement Description:',
+      value: "#{klass.name}Validator::REQUIREMENT_DESCRIPTIONS".safe_constantize
+    }
+    if custom_batch_validator_messages[:value].nil?
+      {}
+    else
+      [custom_batch_validator_messages]
+    end
```
Comment (omgitsbillryan) [inline app/controllers/uploads_controller.rb]: This if-else block doesn't seem to do anything. Am I missing something?

**Review by omgitsbillryan — COMMENTED** [review]
Mind updating the description to be more specific to what changes were made in code? i.e. "added `foo` param to `bar_controller`"

Right now it looks to be a copy+paste from the associated ticket.

File: app/models/upload.rb [hunk 22]
```diff
@@ -148,8 +148,8 @@ def csv_file_headers
   def set_col_sep(first_line)
     self.col_sep = Settings.csv_upload.column_separators
                            .find { |column_separator| first_line.include?(column_separator) }
-
-    error_message = "Unable to determine column separator. #{Upload.valid_col_seps}"
+    valid_col_seps = Upload.valid_col_seps[:value].map { |cs| "\"#{cs}\"" }.join(' and ')
+    error_message = "Unable to determine column separator. #{valid_col_seps}"
```
Comment (omgitsbillryan) [inline app/models/upload.rb]: Does `valid_col_seps` represent valid column separators? The error message language doesn't make this obvious. Maybe something like,

```
Unable to determine column separator. Valid separators = [#{valid_col_seps}].
```

File: app/controllers/uploads_controller.rb [hunk 23]
```diff
@@ -9,42 +9,42 @@ def new
     @upload = Upload.from_csv_type(params[:csv_type])
 
     if @upload.csv_type_check?
-      @requirements = requirements_messages
+      csv_requirements
+
       return
```
Comment (omgitsbillryan) [inline app/controllers/uploads_controller.rb]: Not related to PR necessarily, but I'm trying to better understand.

What does this route return for an http response? Also, what does a blank `return` do from within a controller?

**Review by omgitsbillryan — COMMENTED** [review]
I'm a bit confused. There's a lot of validation happening in the controller that I would expect to see in a validator. Is using vanilla models + validators not an option?

File: app/controllers/uploads_controller.rb (hunk 21 above)
Comment (ksantiagomorgan6) [reply app/controllers/uploads_controller.rb]: That should of been removed, cleaned it up

File: app/controllers/uploads_controller.rb (hunk 23 above)
Comment (ksantiagomorgan6) [reply app/controllers/uploads_controller.rb]: I updated it so it makes more sense with the if statement returning the csv_requirements

**Comment by ksantiagoBAH** [comment]
> I'm a bit confused. There's a lot of validation happening in the controller that I would expect to see in a validator. Is using vanilla models + validators not an option?

I had to use these validators to get the data necessary to display correct info in the views.

File: app/controllers/uploads_controller.rb [hunk 24]
```diff
@@ -109,46 +112,78 @@ def klass
...
+    numericality = { message: 'These columns can only contain numeric values: ', value: [] }
+
+    numericality[:value] = klass_validator(ActiveModel::Validations::NumericalityValidator)
+
+    [numericality] unless numericality[:value].empty?
   end
 
-  def validation_messages
+  def validation_messages_uniqueness
+    uniqueness = { message: 'These columns should contain unique values: ', value: [] }
```
Comment (omgitsbillryan) [inline app/controllers/uploads_controller.rb]: `'These columns should contain unique values: '` kind of looks like "view" code to me. If we were able to move the `message`'s over to our views, we might be able to do away with the arbitrary `{message:, value:}` data structure altogether which would the objects we're passing to our view a bit more straightforward. It would also make a lot of the other code in this controller that's responsible for creating and manipulating this data structure simpler.

**Review by omgitsbillryan — APPROVED** [review]
I don't want to block so, LGTM.

My main objection that this introduces a lot of bloat in the controller that looks like it doesn't belong there. I think a lot of the logic around validations could be moved to the models, and we shouldn't need to create that arbitrary `{message:, value:}` data structure. This along with some comments I made inline that I leave to your discretion.
